File: Algorithms/Permutation_Sequence.cpp
/*
 * Permutation_Sequence.cpp
 * Copyright (C) 2014 fuel <fuel@fuel-3330>
 *
 * Distributed under terms of the MIT license.
 */

#include <iostream>
#include <algorithm>
using namespace std;

class Solution {
public:
    string getPermutation(int n, int k) {
        int circle = 1;
        for(int i=2; i<=n; i++) circle*=i;
        k -= 1; k %= circle;

        string nums(n,'0');
        for(int i=1; i<=n; i++) nums[i-1]+=i;

        string seq;

        while(n){
            circle /= n;
            int idx = k/circle;
            k %= circle;
            seq.push_back(nums[idx]);
            nums.erase(nums.begin()+nums.find(nums[idx]));
            n--;
        }

        return seq;
    }
};

int main() {
    Solution s;   
    for (int k = 1; k<6+1; ++k) {
       cout<<s.getPermutation(3,k)<<endl;
    }
    return 0;
}

File: Algorithms/Permutation_Sequence.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
# vim:fenc=utf-8
#
# Copyright © 2014 fuel <fuel@fuel-3330>
#
# Distributed under terms of the MIT license.
import operator

class Solution:
    # @return a string
    def getPermutation(self, n, k):
        circle = reduce(operator.mul, range(1,n+1))
        k -= 1 # start from 1 not 0, move once
        k %= circle
        nums = range(1,n+1)
        seq = []
        #print circle,k,nums

        while(n):
            circle /= n
            idx = k/circle
            k %= circle
            seq.append(nums[idx])
            nums.remove(nums[idx])
            n -= 1
            #print seq

        return ''.join(map(str,seq))

if __name__=="__main__":
    getPermutation = Solution().getPermutation
    for k in range(1,6+1):
        print getPermutation(3,k)
